pkg/auth/requests/service_account_test.go

```diff
@@ -0,0 +1,348 @@
+package requests
+
+import (
+	"net/http"
+	"reflect"
+	"testing"
+	"time"
+
+	jwtv4 "github.com/golang-jwt/jwt/v4"
+	"github.com/golang/mock/gomock"
+	"github.com/gorilla/mux"
+	v3api "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3/fakes"
+	"github.com/rancher/rancher/pkg/types/config"
+	wranglerfake "github.com/rancher/wrangler/pkg/generic/fake"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	k8sRequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+const (
+	proxyPrefix       = "/k8s/clusters/"
+	tokenReviewSuffix = "/apis/authentication.k8s.io/v1/tokenreviews"
+)
+
+func TestIsTokenExpired(t *testing.T) {
+	tests := []struct {
+		name           string
+		expirationTime *jwtv4.NumericDate
+		want           bool
+	}{
+		{
+			name: "empty expiration",
+			want: false,
+		},
+		{
+			name:           "expired token",
+			expirationTime: jwtv4.NewNumericDate(time.Now().Add(-time.Hour)),
+			want:           true,
+		},
+		{
+			name:           "not expired token",
+			expirationTime: jwtv4.NewNumericDate(time.Now().Add(time.Hour)),
+			want:           false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := isTokenExpired(tt.expirationTime); got != tt.want {
+				t.Errorf("isTokenExpired() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestAuthenticate(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name                  string
+		startUser             string
+		expectedUser          string
+		settingEnabled        bool
+		token                 string
+		tokenReviewAuthStatus bool
+		downstreamRequest     *http.Request
+		clusterID             string
+		isAuthenticated       bool
+		expectedError         bool
+		omitProxyConfig       bool
+	}{
+		{
+			name: "everything valid auth succeeds secure mode",
+			token: makeJWT(jwtv4.MapClaims{
+				"sub": "system:serviceaccount:vault:token-reviewer",
+				"exp": time.Now().Add(time.Hour).Unix(),
+			}),
+			startUser:             "system:cattle:error",
+			expectedUser:          "expected-username",
+			settingEnabled:        true,
+			tokenReviewAuthStatus: true,
+			downstreamRequest:     generateDownstreamRequest("POST", "testclusterid"+tokenReviewSuffix),
+			clusterID:             "testclusterid",
+			isAuthenticated:       true,
+			expectedError:         false,
```

Comment:
It defaults to false so you could easily leave these ones out?

Reply:
could leave it out, but I sort of like being explicit.